egress/src/reports/idf_event.rs

```diff
@@ -0,0 +1,378 @@
+        }
+    }
+
+    if unit == IdfUnit::Lsha {
+        maximum.intensity = mm_to_lsha(maximum.intensity, duration)
+    }
+
+    maximum
+}
+
+// TODO: should spawn in separate thread and use par_iter instead of into_iter?
+fn collect_idf_events(durations: &[u32], data: Vec<RainfallDatum>, unit: IdfUnit) -> Vec<IdfEvent> {
+    durations
+        .iter()
+        .map(|&duration| calculate_idf_event(duration, &data, unit))
+        .collect()
+}
+
+pub async fn idf_event_handler(
+    Path(station_id): Path<i32>,
+    State(pools): State<DbPools>,
+    State(tables): State<PatchworkTables>,
+    Query(params): Query<IdfEventParams>,
+    Extension(roles): Extension<Option<Vec<i32>>>,
+) -> Result<Json<IdfEventResp>, (StatusCode, String)> {
+    let idf_event_label = PatchworkLabel::new(
+        station_id,
+        PRECIPITATION_PARAM_ID,
+        DEFAULT_LEVEL,
+        DEFAULT_SENSOR,
+    );
+
+    let (patches, pool) = patchwork::get_patches(
+        params.fromtime,
+        params.totime,
+        idf_event_label,
+        tables,
+        pools,
+        roles,
+    )
+    .await
+    .map_err(internal_error)?;
+
+    if patches.is_empty() {
+        return Err((
+            StatusCode::NOT_FOUND,
+            "No applicable timeseries in the given time period".to_string(),
+        ));
+    };
+
+    // TODO: this should be handled by auth
+    let conn = pool.get().await.map_err(internal_error)?;
+
+    let data = fetch_rain_data(patches, &conn)
+        .await
+        .map_err(internal_error)?;
+
+    if data.is_empty() {
+        return Err((
+            StatusCode::NOT_FOUND,
+            format!("No precipitation data found for station {station_id}"),
+        ));
+    }
+
+    // We allow any provided duration that is less or equal to `MAX_ALLOWED_DURATION`
+    let inputs: Option<Vec<u32>> = params.durations.map(|values| {
+        values
+            .into_iter()
+            .filter(|val| *val <= MAX_ALLOWED_DURATION)
+            .collect()
+    });
+
+    let durations = inputs.as_deref().unwrap_or(DEFAULT_DURATIONS);
+    let values = collect_idf_events(durations, data, params.unit);
+
+    Ok(Json(IdfEventResp {
+        station_id,
+        unit: params.unit,
+        values,
+    }))
+}
+
+/// Response struct returned by the availability endpoint
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct IdfEventAvailability {
+    /// List of stations for which IDF calculation is possible
+    pub stations: Vec<i32>,
+}
+
+fn is_idf_event_timeseries(label: &PatchworkLabel) -> bool {
+    label.param_id == PRECIPITATION_PARAM_ID
+        && label.level == DEFAULT_LEVEL
+        && label.sensor == DEFAULT_SENSOR
+}
+
+pub async fn idf_event_availability_handler(
+    State(patchwork_tables): State<PatchworkTables>,
+) -> Result<Json<IdfEventAvailability>, (StatusCode, String)> {
+    // TODO: need to implement this also for restricted?
+    let ot = patchwork_tables.open.read().map_err(internal_error)?;
```

Comment:
Need to pass roles here to also deliver restricted? 
Will have same weakness as patchwork avaiable, in that if you are allowed to see any restricted ts you will see all of them listed here (but you cannot get the data). 
We can try to filter by roles here, should also implement same filtering in patchwork available.

Reply:
Done